contracts/V1/StateChannelDiamondProxy/JoinChannelFacet.sol

```
@@ -0,0 +1,152 @@
+     * @notice Joins participants to the channel by submitting a complete join channel request
+     *
+     * @param channelId The channel identifier
+     * @param joinChannelConfirmations Array of join channel confirmations to process
+     * @param dispute  to create new fork commitment
+     */
+    function joinChannel(
+        bytes32 channelId,
+        JoinChannelConfirmation[] memory joinChannelConfirmations,
+        Dispute memory dispute
+    ) external {
+        if (!isSnapshotLatest(channelId)) {
+            revert ErrorSnapshotNotLatest();
+        }
+
+        // SETUP PHASE
+        // make a list of join channels
+        JoinChannel[] memory joinChannels = new JoinChannel[](joinChannelConfirmations.length);
+        for (uint256 i = 0; i < joinChannelConfirmations.length; i++) {
+            joinChannels[i] =
+                abi.decode(joinChannelConfirmations[i].signedJoinChannel.encodedJoinChannel, (JoinChannel));
+        }
+        // make a list of joining participants
+        address[] memory joiningParticipants = new address[](joinChannels.length);
+        for (uint256 i = 0; i < joinChannels.length; i++) {
+            joiningParticipants[i] = joinChannels[i].participant;
+        }
+
+        // Create validity tracking arrays
+        bool[] memory isValid = new bool[](joinChannelConfirmations.length);
+        uint256 validCount = 0;
+
+        // VALIDATION PHASE - Filter out invalid join channels
+
+        // 1. Filter by join channel deadlines
+        for (uint256 i = 0; i < joinChannels.length; i++) {
+            if (joinChannels[i].deadlineTimestamp > block.timestamp) {
+                isValid[i] = true;
+                validCount++;
+            }
+        }
+
+        // 2. Filter by threshold signatures from existing participants
+
+        address[] memory allParticipants = StateChannelUtilLibrary.concatenateWithoutDuplicates(
+            getSnapshotParticipants(channelId), getPendingParticipants(channelId)
```

pendingParticipants will be here empty since the deposits are not yet processed. Imo since in the end all of this needs to be atomic (we can't do partial failure as I wanted), since the `dispute` commits to the specific `latestJoinChannelBlockHash` and the dispute happens in the same transaction so it has to assume all joins have ran successfully (if a single join fails the hash won't match and the dispute will fail before being committed). Best to verify the signatures at the end when all deposits are processed, since then it's a clean interface - just need to verify `snapshot.participants` + `pendingParticpants` that contains everyone - so we batch all the signatures together and just call verifyThreshold once in the end 

Ideally later we find a way where it doesn't have to be atomic and we can account for partial failures

---

ok I guess i had a terminology misunderstanding

for you the pending participants are the ones that **just joined**

for me pending participants are from previous rounds of joining, and yes it would be empty since we just updated the state snapshot, which also clears the pending

for me the mental image was
- current snapshot participants
- pending (should be empty indeed)
- new comers

for you the image is 
- current snapshot participants
- new joiners => becoming the pending
- verify threshold with the pending aka the new joiners


still, for the sake of quickness and gas efficiency, isn't it better to run as many validation as possible early and fail early?

---

Yes and yes, but the signature verification is deterministic - the tx submitter knows are the sigantures valid or not before submitting, what they don't know will the deposits be processed successfully so it's best to fail early on that and on 'time' (is it expired) if all of that succeeds, the rest (signature verification) will resolve deterministically - so for an 'honest' submitter verifying signatures last saves them gas on average, since that's not the part that will fail 

We can discuss today in the call